core/domain/stats_jobs_one_off.py

```diff
@@ -28,14 +34,248 @@
+                       'event_type': feconf.EVENT_TYPE_STATE_COMPLETED,
+                       'version': item.exp_version,
+                       'state_name': item.state_name
+                   })
+        elif isinstance(
+                item, stats_models.AnswerSubmittedEventLogEntryModel):
+            yield (item.exp_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_ANSWER_SUBMITTED,
+                       'version': item.exp_version,
+                       'state_name': item.state_name,
+                       'is_feedback_useful': item.is_feedback_useful
+                   })
+        elif isinstance(item, stats_models.StateHitEventLogEntryModel):
+            yield (item.exploration_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_STATE_HIT,
+                       'version': item.exploration_version,
+                       'state_name': item.state_name,
+                       'session_id': item.session_id
+                   })
+        elif isinstance(item, stats_models.SolutionHitEventLogEntryModel):
+            yield (item.exp_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_SOLUTION_HIT,
+                       'version': item.exp_version,
+                       'state_name': item.state_name,
+                       'session_id': item.session_id
+                   })
+        elif isinstance(
+                item,
+                stats_models.ExplorationActualStartEventLogEntryModel):
+            yield (item.exp_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION,
+                       'version': item.exp_version,
+                       'state_name': item.state_name
+                   })
+        elif isinstance(
+                item, stats_models.CompleteExplorationEventLogEntryModel):
+            yield (item.exploration_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_COMPLETE_EXPLORATION,
+                       'version': item.exploration_version,
+                       'state_name': item.state_name
+                   })
+
+
+    @staticmethod
+    def reduce(exp_id, values):
+        values = map(ast.literal_eval, values)
+        values = sorted(values, key=lambda x: x['version'])
+        versions = [x['version'] for x in values]
+        # Get a copy of the uncorrupted *_v1 statistics
+        old_stats = stats_services.get_multiple_exploration_stats_by_version(
+            exp_id, versions)
+        # Get list of snapshot models for each version of the exploration
+        snapshots_by_version = (
+            exp_models.ExplorationModel.get_snapshots_metadata(
+                exp_id, versions))
+
+        exp_stats_dicts = []
+        for version, events in itertools.groupby(values,
+                                                 key=lambda x: x['version']):
+            v1_stats = old_stats[version-1]
```

Comment:
I think v1 is ambiguous since we're also dealing with exploration versions (and in any case the stuff in old_stats includes some v2 fields). Maybe call it datastore_stats_for_this_version or something like that.